### PR TITLE
Add on event

### LINKED ISSUE #20: Add new onEvent parameter to MonoConnect constructor
**Context**
The MonoConnect widget provides basic callback functions such as onLoad, onClose, and onSuccess on instantiation.

**Problem**
Developers do not have insight into the widget's lifecycle for debugging. There is also no error reporting.

**Solution**
A new onEvent parameter will allow us to publish all events from the widget to one place.

Events include:

- Loaded
- Opened
- Closed
- Error
- Success

onEvent will return two parameters (eventName: string, metadata: object)

{
error_code // [Invalid Credentials, Invalid MFA, Invalid App Key]
 error_message // A more detailed message based on error_code
 exit_status  // [connected, failed, cancelled]
code 
 timestamp
view_name
}

We will continue to support onSuccess, onClose, and onLoad.

### PR DESCRIPTION
**Pull request by tristan-mono**
Added new optional onEvent function to Connect class. Returns parameters eventName: String, metadata: Object

eventName can be on of ['OPENED', 'CLOSED', 'LOADED', 'SUCCESS']
metadata returns an object with relevant information: 
{
code: String or null // the auth code to return to 
error_code: null // does not return anything for now, we can add this later if possible I'd like to pass more data from the connect.mono.co widget
" error_message": null // ditto
" timestamp": 1622225295275 // unix timestamp from epoch
}

### CODE REVIEW

File: connect.js [hunk 1]
```diff
@@ -25,15 +26,18 @@ function connect({
     onSuccess = arguments[1].onSuccess || isRequired("onSuccess callback");
     onLoad = arguments[1].onLoad || anonFunc;
     rest = {};
+    onEvent = arguments[1].onEvent || anonFunc;
```
Comment (umar-abdullahi) [inline connect.js]: this block handles a deprecated Mono connect constructor, so won't support new features

File: connect.js [hunk 2]
```diff
@@ -86,10 +94,11 @@ connect.prototype.close = function () {
   window.removeEventListener("message", this.eventHandler, false);
   connect.prototype.utils.closeWidget();
   this.onClose();
+  this.onEvent('CLOSED', connect.prototype.utils.metadata({}));
```
Comment (umar-abdullahi) [inline connect.js]: closed event would now receive the `pageName` where the widget was closed, this would be sent form the widget

File: utils.js [hunk 3]
```diff
@@ -67,17 +68,19 @@ var utils = () => {
...
     var container = document.getElementById("mono-connect--widget-div");
     var loader = document.getElementById("mono-connect-app-loader");
     var frame = document.getElementById("mono-connect--frame-id");
     container.style.visibility = "visible";
     container.style.display = "flex";
     loader.style.display = "block";
 
-    setTimeout(() => { 
-      turnOnVisibility(); 
+    onEvent('OPENED', metadata({}))
```
Comment (umar-abdullahi) [inline utils.js]: opened event has to be from connect, that way we are sure the widget was opened, also would passed reference if available

File: utils.js [hunk 4]
```diff
@@ -109,7 +112,35 @@ var utils = () => {
     document.head.appendChild(styleSheet);
   }
 
+  function metadata(data){
```
Comment (umar-abdullahi) [inline utils.js]: we don't need this function, all metadata would be passed from connect

File: utils.js (hunk 3 above)
Comment (tristan-mono) [reply utils.js]: There is no event for open being published to the `message` event. Is this feature still coming?

File: utils.js (hunk 4 above)
Comment (tristan-mono) [reply utils.js]: Are you referring to the data object returned as part of the event? If so should I just pass that back to the onEvent callback then?

File: connect.js (hunk 2 above)
Comment (tristan-mono) [reply connect.js]: Currently not receiving a pageName in event.data

File: README.md [hunk 5]
```diff
@@ -145,11 +155,77 @@ This method programatically hides the widget after it's been opened.
...
+  }
+});
+
+```
+
+#### <a name="eventName"></a> `eventName`
+
+Event names corespond to the `type` key returned by the raw event data. Possible options are in the table below.
+
+| Raw Event Type | eventName |
```
Comment (umar-abdullahi) [inline README.md]: We don't have to document the raw event type, just the eventName and data object properties of each event

**Pull request merged**